### DISCUSSION TITLE
`Base.metadata.create_all` doesn't create all tables with `create_async_engine`

### DISCUSSION THREAD

**Discussion by wiseaidev** [sqlalchemy/sqlalchemy#8650] [Usage Questions] [answered]
### Describe the bug

Hey everyone. I am currently using `sqlalchemy.ext.asyncio.create_async_engine` with [`databases[aiomysql]`](https://github.com/encode/databases) extension to create a `mysql` database. However, the `sqlalchemy.engine.Engine` triggers only one table to be created and not all data tables. The following are all the major parts of the SQLAlchemy configs:

- [`init_models`](https://github.com/wiseaidev/fastapi-singlestore-backend/blob/b6321a3e19ae3cbf9ede107f1d3a8d42c712b0a3/app/utils/session.py#L30-L46).
- [mixins file](https://github.com/wiseaidev/fastapi-singlestore-backend/blob/main/app/utils/mixins.py).
- [`users` table](https://github.com/wiseaidev/fastapi-singlestore-backend/blob/main/app/users/model.py).
- [tables that are not created at the startup](https://github.com/wiseaidev/fastapi-singlestore-backend/blob/main/app/auth/model.py).
- The `init_models` is being [called here](https://github.com/wiseaidev/fastapi-singlestore-backend/blob/b6321a3e19ae3cbf9ede107f1d3a8d42c712b0a3/app/main.py#L62).

Not sure what i am missing here. It could be a `databases` issue or `sqlalchemy` issue. Not quite sure. But I think it is more likely the latter.

### To Reproduce

Assuming you have `mysql server` configured on your machine, clone [this repo](https://github.com/wiseaidev/fastapi-singlestore-backend):

```sh

git clone https://github.com/wiseaidev/fastapi-singlestore-backend.git
cd fastapi-singlestore-backend
make install
make run
```
Not that you have to change [this line of code](https://github.com/wiseaidev/fastapi-singlestore-backend/blob/b6321a3e19ae3cbf9ede107f1d3a8d42c712b0a3/app/utils/session.py#L24) according to your mysql client config.

### Error

Not actually an error, but rather an incorrect SQLAlchemy engine behavior.

### Versions

- OS: Ubuntu 20.04.4 LTS
- Python: 3.9.10
- SQLAlchemy: 1.4.41
- Database: MySQL
- DBAPI (eg: psycopg, cx_oracle, mysqlclient): [databases](https://github.com/encode/databases)

### Additional context

Setting `echo=True` in the `create_async_engine` results in the following being logged into the terminal:

<details>
<summary><code>create_async_engine logs</code></summary>

```sh
2022-10-17 12:50:01,756 INFO sqlalchemy.engine.Engine SELECT DATABASE()
2022-10-17 12:50:01,756 INFO sqlalchemy.engine.Engine [raw sql] ()
2022-10-17 12:50:02,224 INFO sqlalchemy.engine.Engine SELECT @@sql_mode
2022-10-17 12:50:02,224 INFO sqlalchemy.engine.Engine [raw sql] ()
2022-10-17 12:50:02,451 INFO sqlalchemy.engine.Engine SELECT @@lower_case_table_names
2022-10-17 12:50:02,452 INFO sqlalchemy.engine.Engine [raw sql] ()
2022-10-17 12:50:02,903 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2022-10-17 12:50:02,904 INFO sqlalchemy.engine.Engine SELECT COUNT(*) FROM information_schema.tables WHERE table_schema = %s AND table_name = %s
2022-10-17 12:50:02,905 INFO sqlalchemy.engine.Engine [generated in 0.00037s] ('chat', 'users')
2022-10-17 12:50:03,130 INFO sqlalchemy.engine.Engine 
DROP TABLE users
2022-10-17 12:50:03,130 INFO sqlalchemy.engine.Engine [no key 0.00027s] ()
2022-10-17 12:50:03,415 INFO sqlalchemy.engine.Engine SELECT COUNT(*) FROM information_schema.tables WHERE table_schema = %s AND table_name = %s
2022-10-17 12:50:03,415 INFO sqlalchemy.engine.Engine [cached since 0.5109s ago] ('chat', 'users')
2022-10-17 12:50:03,641 INFO sqlalchemy.engine.Engine 
CREATE TABLE users (
	id BIGINT NOT NULL AUTO_INCREMENT, 
	creation_date DATETIME, 
	modified_date DATETIME, 
	first_name VARCHAR(20), 
	last_name VARCHAR(20), 
	email VARCHAR(50), 
	password VARCHAR(120), 
	phone_number VARCHAR(20), 
	bio VARCHAR(60), 
	profile_picture VARCHAR(220), 
	chat_status VARCHAR(20), 
	user_status INTEGER, 
	user_role VARCHAR(20), 
	PRIMARY KEY (id)
)ENGINE=InnoDB


2022-10-17 12:50:03,641 INFO sqlalchemy.engine.Engine [no key 0.00029s] ()
/home/harmouch/Desktop/SingleStore Hackathon/fastapi-singlestore-backend/.venv/lib/python3.9/site-packages/aiomysql/cursors.py:239: Warning: Feature 'storage engine plugins' is not supported by SingleStore. Execution will continue, but the feature will be ignored.
  await self._query(query)
2022-10-17 12:50:04,146 INFO sqlalchemy.engine.Engine CREATE INDEX ix_users_password ON users (password)
2022-10-17 12:50:04,146 INFO sqlalchemy.engine.Engine [no key 0.00026s] ()
2022-10-17 12:50:05,528 INFO sqlalchemy.engine.Engine CREATE INDEX ix_users_email ON users (email)
2022-10-17 12:50:05,529 INFO sqlalchemy.engine.Engine [no key 0.00028s] ()
2022-10-17 12:50:06,818 INFO sqlalchemy.engine.Engine CREATE INDEX ix_users_user_status ON users (user_status)
2022-10-17 12:50:06,818 INFO sqlalchemy.engine.Engine [no key 0.00027s] ()
2022-10-17 12:50:07,926 INFO sqlalchemy.engine.Engine COMMIT
2022-10-17 12:50:08,382 INFO sqlalchemy.pool.impl.AsyncAdaptedQueuePool Pool disposed. Pool size: 30  Connections in pool: 0 Current Overflow: -30 Current Checked out connections: 0
```

</details>

As you can see only the `users` table is being created. `access_tokens` and `black_listed_tokens` are not created. At the time being, a workaround i can think of is to manually create these tables in `mysql`. However, this process is quite tidious. It would be nice if `SQLalchemy` can do it for me.

**Comment by CaselIT** [comment, marked as answer]
Hi,

This is the expected behaviour and it's not something that sqlalchemy can really do anything about.
You need to ensure that all models/tables that are defined in your application are imported before running create_all. If a module is not imported its table is not registered in the metadata, so sqlalchemy cannot create it.

The usual way is to have a package with the models/tables, and import all of the m in the `__init__` file, then import that package before the `create_all` call

**Reply by wiseaidev** [reply]
Hey @CaselIT,

Thanks for your response. It is truly helpful. Actually, I wasn't aware of how the engine works until I realized that only the `users` table is being imported into another module. The other two tables weren't imported anywhere.

Having solved this issue, now i am getting a vendor specific(SingleStore) issue:

```sh
sqlalchemy.exc.OperationalError: (pymysql.err.OperationalError) (1706, "Forwarding Error (node-XXXXXXX): Feature 'FOREIGN KEY on COLUMNAR table' is not supported by SingleStore.")
```

I have no idea what that means. For reference, this is what the tables look like after sqlalchemy translated them:

```SQL
CREATE TABLE users (
	id BIGINT NOT NULL AUTO_INCREMENT, 
	creation_date DATETIME, 
	modified_date DATETIME, 
	first_name VARCHAR(20), 
	last_name VARCHAR(20), 
	email VARCHAR(50), 
	password VARCHAR(120), 
	phone_number VARCHAR(20), 
	bio VARCHAR(60), 
	profile_picture VARCHAR(220), 
	chat_status VARCHAR(20), 
	user_status INTEGER, 
	user_role VARCHAR(20), 
	PRIMARY KEY (id)
)ENGINE=InnoDB

CREATE TABLE access_tokens (
	id BIGINT NOT NULL AUTO_INCREMENT, 
	creation_date DATETIME, 
	modified_date DATETIME, 
	user BIGINT, 
	token VARCHAR(120), 
	PRIMARY KEY (id), 
	FOREIGN KEY(user) REFERENCES users (id)
)ENGINE=InnoDB
```

Any idea? Thanks for your consideration.

**Reply by wiseaidev** [reply]
Hey, @CaselIT, Do you have any idea how to tell SQLAlchemy to translate the `access_token` table into the following:

```SQL
CREATE ROWSTORE REFERENCE TABLE access_tokens (
    id BIGINT NOT NULL AUTO_INCREMENT, 
    creation_date DATETIME, 
    modified_date DATETIME, 
    user BIGINT, 
    token VARCHAR(120), 
    PRIMARY KEY (id), 
    FOREIGN KEY(user) REFERENCES users (id)
)ENGINE=InnoDB
```

This is the only way that this table can be created in SIngleStore after reading [this post](https://dev.to/franckpachot/singlestore-compared-to-yugabytedb-jfp). I just need to append `ROWSTORE REFERENCE` keywords to the `CREATE` command. How can you do this in SQLAlchemy?

**Reply by CaselIT** [reply]
I believe you can use [prefixes](https://docs.sqlalchemy.org/en/14/core/metadata.html?highlight=prefix#sqlalchemy.schema.Table.params.prefixes) for that

**Reply by wiseaidev** [reply]
Hey @CaselIT, thank you very much for providing a helpful response. However, I was having difficulty figuring out where to put these values because i am using a declarative approach to define my models. For future reference, you have to place these values in the `__table_args__` attribute:

```python
__table_args__ = {"mysql_engine": "InnoDB", "prefixes": ["ROWSTORE", "REFERENCE"]}
```

Thank you again for your help!

**Reply by CaselIT** [reply]
@zzzeek we are missing `__table_args__ ` and `__mapper_args__ ` from the `DeclarativeBase` typing

**Reply by zzzeek** [reply]
I dont recall if I had a reason for that, given that ``__tablename__`` is broken already w/ the classvar thing
